### PR TITLE
Make task check period configurable in workflows

### PR DESCRIPTION
**Pull request by YuanTingHsieh**
- Make task check period configurable in the following workflows:
  - nvflare/app_common/workflows/cross_site_model_eval.py
  - nvflare/app_common/workflows/cyclic_ctl.py
  - nvflare/app_common/workflows/scatter_and_gather.py
  - nvflare/app_common/workflows/scatter_and_gather_scaffold.py

- Added "order" argument to `CyclicController`

### CODE REVIEW

File: nvflare/app_common/workflows/scatter_and_gather_scaffold.py [hunk 1]
```diff
@@ -189,13 +204,22 @@ def control_flow(self, abort_signal: Signal, fl_ctx: FLContext) -> None:
...
+                    self.persistor.save(self._global_weights, fl_ctx)
+                    self.fire_event(AppEventType.AFTER_LEARNABLE_PERSIST, fl_ctx)
+                    self.log_info(fl_ctx, "End persist model on server.")
 
                 self.fire_event(AppEventType.ROUND_DONE, fl_ctx)
                 self.log_info(fl_ctx, f"Round {self._current_round} finished.")
 
+                self._current_round += 1
+
+                if self._snapshot_every_n_rounds != 0 and self._current_round % self._snapshot_every_n_rounds == 0:
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/app_common/workflows/scatter_and_gather_scaffold.py:219]: persist/save entire system snapshot is different VS persist/save a global model, they should use different variable.

File: nvflare/app_common/workflows/cyclic_ctl.py [hunk 2]
```diff
@@ -139,7 +146,9 @@ def control_flow(self, abort_signal: Signal, fl_ctx: FLContext):
                 )
                 self.persistor.save(self._last_learnable, fl_ctx)
                 self.log_debug(fl_ctx, "Ending current round={}.".format(self._current_round))
-                self._engine.persist_components(fl_ctx, completed=False)
+                if self._snapshot_every_n_rounds != 0 and self._current_round % self._snapshot_every_n_rounds == 0:
```
Comment (yanchengnv) [inline nvflare/app_common/workflows/cyclic_ctl.py:149]: To be consistent with SAG, should we add another arg for persist_every_n_rounds to control self.persistor.save frequency?

File: nvflare/app_common/workflows/scatter_and_gather_scaffold.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/app_common/workflows/scatter_and_gather_scaffold.py:219]: I see a lot of repeat of the same code as the base class SAG. I took a closer look at this controller. It's not written properly. It's not even needed - all the differences between it and the SAG should be put into a different Shareable Generator.

Let's do this later.

**Review by yanchengnv — CHANGES_REQUESTED** [review]
Please make the cyclic controller consistent with SAG - control how frequent the persistor.save is called.

File: nvflare/app_common/workflows/cyclic_ctl.py (hunk 2 above)
Comment (YuanTingHsieh) [reply nvflare/app_common/workflows/cyclic_ctl.py:161]: thanks, fixed

File: nvflare/app_common/workflows/cyclic_ctl.py [hunk 3]
```diff
@@ -137,9 +149,18 @@ def control_flow(self, abort_signal: Signal, fl_ctx: FLContext):
...
+                    self.fire_event(AppEventType.BEFORE_LEARNABLE_PERSIST, fl_ctx)
+                    self.persistor.save(self._last_learnable, fl_ctx)
+                    self.fire_event(AppEventType.AFTER_LEARNABLE_PERSIST, fl_ctx)
+                    self.log_info(fl_ctx, "End persist model on server.")
+
                 self.log_debug(fl_ctx, "Ending current round={}.".format(self._current_round))
-                self._engine.persist_components(fl_ctx, completed=False)
+                if self._snapshot_every_n_rounds != 0 and self._current_round % self._snapshot_every_n_rounds == 0:
+                    # Call the self._engine to persist the snapshot of all the FLComponents
+                    self._engine.persist_components(fl_ctx, completed=False)
```
Comment (yanchengnv) [inline nvflare/app_common/workflows/cyclic_ctl.py:163]: Where is self._current_round advanced (+1)?

File: nvflare/app_common/workflows/cyclic_ctl.py [hunk 4]
```diff
@@ -33,6 +33,9 @@ def __init__(
         persistor_id="persistor",
         shareable_generator_id="shareable_generator",
         task_name="train",
+        task_check_period: float = 0.5,
+        persist_every_n_rounds: int = 1,
+        snapshot_every_n_rounds: int = 1,
     ):
```
Comment (yanchengnv) [inline nvflare/app_common/workflows/cyclic_ctl.py:39]: Need another arg for relay order:
Fixed - use the same order for all rounds
Random - randomize client order for each round
Random_No_Consecutive_Clients - randomize client order for each round AND not same client twice in a row (last of the last round and first of next round must be different)

Default is FIXED.

**Review by yanchengnv — CHANGES_REQUESTED** [review]

File: nvflare/app_common/workflows/cyclic_ctl.py (hunk 3 above)
Comment (YuanTingHsieh) [reply nvflare/app_common/workflows/cyclic_ctl.py:163]: this controller use a for loop so it advance (+1) in the for loop

File: nvflare/app_common/workflows/cyclic_ctl.py (hunk 4 above)
Comment (YuanTingHsieh) [reply nvflare/app_common/workflows/cyclic_ctl.py:49]: thanks, fixed

**Review by yanchengnv — CHANGES_REQUESTED** [review]
There is a bug in using self._engine in cyclic_ctl.py.
Please make sure to test it before submitting the PR.

File: nvflare/app_common/workflows/cyclic_ctl.py [hunk 5]
```diff
@@ -89,6 +122,22 @@ def start_controller(self, fl_ctx: FLContext):
         fl_ctx.set_prop(AppConstants.GLOBAL_MODEL, self._last_learnable, private=True, sticky=True)
         self.fire_event(AppEventType.INITIAL_MODEL_LOADED, fl_ctx)
 
+        self._participating_clients = self._engine.get_clients()
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/app_common/workflows/cyclic_ctl.py:125]: self._engine was set in the base class: https://github.com/NVIDIA/NVFlare/blob/dev/nvflare/apis/impl/controller.py#L92

[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/app_common/workflows/cyclic_ctl.py:125]: Thanks for the review. I've tested this with hello-cyclic and it is working.

File: nvflare/app_common/workflows/cyclic_ctl.py [hunk 6]
```diff
@@ -44,11 +57,22 @@ def __init__(
                 Defaults to "persistor".
             shareable_generator_id (str, optional): id of shareable generator. Defaults to "shareable_generator".
             task_name (str, optional): the task name that clients know how to handle. Defaults to "train".
+            task_check_period (float, optional): interval for checking status of tasks. Defaults to 0.5.
+            persist_every_n_rounds (int, optional): persist the global model every n rounds. Defaults to 1.
+                If n is 0 then no persist.
+            snapshot_every_n_rounds (int, optional): persist the server state every n rounds. Defaults to 1.
+                If n is 0 then no persist.
+            order (str, optional): the order of relay.
+                If FIXED means everytime it is the same order (ordered by client names).
```
[the remark this answers is not in the document]
Comment (YuanTingHsieh) [reply nvflare/app_common/workflows/cyclic_ctl.py]: Thanks I've changed the text.

File: nvflare/app_common/workflows/cyclic_ctl.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (yanchengnv) [reply nvflare/app_common/workflows/cyclic_ctl.py:125]: Oh that's why the base class Controller already set self._engine in initialize_run().

**Review by yanchengnv — APPROVED** [review]
LGTM

**Pull request merged**